inkcut/core/svg.py

```diff
@@ -665,24 +638,57 @@ def __init__(self, e, ids=None):
         if self.isParentSvg:
             self._doc = etree.parse(e)
             self._svg = self._doc.getroot()
-            self.viewBox = QtCore.QRectF(0, 0, -1, -1)
-        
-        super(QtSvgDoc, self).__init__(self._svg)
-    
+            e = self._svg
+            box = e.attrib.get('viewBox')
+            if box is None:
+                self.viewBox = None
+            else:
+                (x, y, width, height) = box.split()
+                # Before setting the viewbox, so no unit = px:
+                self.width = self.parseUnit(e.attrib.get('width',None))
+                self.height = self.parseUnit(e.attrib.get('height',None))
+                self.viewBox = QtCore.QRectF(
+                    float(x), float(y),
+                    float(width), float(height))
+
+        super(QtSvgDoc, self).__init__(self, self._svg)
+
     def parseTransform(self, e):
         t = QtGui.QTransform()
-        # transforms don't apply to the root svg element
+        # x/y transforms don't apply to the root svg element
         if not self.isParentSvg:
             x, y = map(self.parseUnit, (e.attrib.get('x', 0),
                                         e.attrib.get('y', 0)))
             t.translate(x, y)
-            
-        # TODO: Handle width/height and viewBox stuff
-        #         else:
-        #             w,h = map(self.parseUnit,(e.attrib.get('width',None),
-        # e.attrib.get('height',None)))
-        #             if w is not None or h is not None:
-        #                 sx,sy = 
-        #             
-            
+
         return t
+
+    def numericValue(self, value):
+        if isinstance(value, (int, float)):
+            return value
+        try:
+            return int(value)
+        except ValueError:
+            try:
+                return float(value)
+            except ValueError:
+                return None
+
+    # Cannot be static because the size of a unitless value depends
+    # on the document viewport and viewbox
+    def parseUnit(self, value):
+        """ Returns userunits given a string representation of units
+        in another system
+        """
+        numericValue = self.numericValue(value)
+        if numericValue is None:
+            return parse_unit(value)
+        else:
+            if self.viewBox is None:
+                # Then these are pixels. Do they need conversion?
+                return numericValue
+            else:
+                # TODO in theory the scaling depends on the 'preserveAspectRatio'
+                # and perhaps even on whether this is an x or an y distance (or even angled?)
+                # For now assume the viewBox maintains the image aspect ratio.
+                return numericValue * self.width / self.viewBox.width()
```

Comment:
If I'm understanding this correctly, the viewBox is not a transform that applies to all child nodes, but dictates only how 'unitless' numeric values are to be interpreted.

For example, if we have an SVG document with `width="10cm" height="10cm" viewBox="0 0 10 10"`, a `rect width="1" height="1"` will be 1cm by 1cm, and a `rect width="2cm" height="2cm"` is 2cm by 2cm.

However, in an SVG document with `width="20cm" height="20cm" viewBox="0 0 10 10"`, a `rect width="1" height="1"` will be 2cm by 2cm, while a `rect width="2cm" height="2cm"` is still 2cm by 2cm.

So we can't just add a transform to the whole document since the viewbox should impact the elements that had no units, but not those that did have units.

Clearly my understanding of this is incomplete, however, as it doesn't render all my test documents correctly yet. Perhaps the transforms should affect the viewbox as well?